nifi-toolkit/nifi-toolkit-zookeeper-migrator/src/main/java/org/apache/nifi/toolkit/zkmigrator/ZooKeeperMigrator.java

```diff
@@ -284,8 +287,29 @@ private Stat transmitNode(ZooKeeper zooKeeper, DataStatAclNode node) {
     }
 
     private ZooKeeper getZooKeeper(ZooKeeperEndpointConfig zooKeeperEndpointConfig, AuthMode authMode, byte[] authData) throws IOException {
+        CountDownLatch connectionLatch = new CountDownLatch(1);
         ZooKeeper zooKeeper = new ZooKeeper(zooKeeperEndpointConfig.getConnectString(), 3000, watchedEvent -> {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("ZooKeeper server state changed to {} in {}", watchedEvent.getState(), zooKeeperEndpointConfig);
+            }
+            if (watchedEvent.getType().equals(Watcher.Event.EventType.None) && watchedEvent.getState().equals(Watcher.Event.KeeperState.SyncConnected)) {
+                connectionLatch.countDown();
+            }
         });
+
+        final boolean connected;
+        try {
+            connected = connectionLatch.await(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException(String.format("interrupted while waiting for ZooKeeper connection to %s", zooKeeperEndpointConfig), e);
+        }
+
+        if (!connected) {
+            throw new IOException(String.format("unable to connect to %s, state is %s", zooKeeperEndpointConfig, zooKeeper.getState()));
```

Comment:
I don't think it's necessary, since the migrator will exit if a connection with ZK cannot be established, but I will add it for the sake of explicit cleanup.